finality-aleph/src/aggregator/mod.rs

```diff
@@ -0,0 +1,150 @@
+use aleph_bft::{MultiKeychain, Recipient, Signable};
+use codec::Codec;
+use futures::{channel::mpsc, StreamExt};
+use log::{debug, trace, warn};
+use std::{
+    collections::{HashMap, HashSet, VecDeque},
+    fmt::Debug,
+    marker::PhantomData,
+};
+
+mod multicast;
+
+pub use multicast::{Hash, Multicast, Multisigned, NetworkData as RmcNetworkData, SignableHash};
+
+/// A type encapsulating three different results of the `process_network_messages` method
+pub(crate) enum AggregatorError {
+    NetworkChannelClosed,
+}
+
+/// A wrapper around an `rmc::Multicast` returning the signed hashes in the order of the [`Multicast::start_multicast`] calls.
+pub(crate) struct BlockSignatureAggregator<
+    'a,
+    H: Hash + Copy,
+    D: Clone + Codec + Debug + Send + Sync + 'static,
+    N: DataNetwork<D>,
+    MK: MultiKeychain,
+    RMC: Multicast<H>,
+> {
+    messages_for_rmc: mpsc::UnboundedSender<D>,
+    messages_from_rmc: mpsc::UnboundedReceiver<D>,
+    signatures: HashMap<H, MK::PartialMultisignature>,
+    hash_queue: VecDeque<H>,
+    network: N,
+    multicast: RMC,
+    last_hash_placed: bool,
+    started_hashes: HashSet<H>,
+    metrics: Option<Metrics<H>>,
+    marker: PhantomData<&'a H>,
+}
+
+impl<
+        'a,
+        H: Copy + Hash,
+        D: Clone + Codec + Debug + Send + Sync,
+        N: DataNetwork<D>,
+        MK: MultiKeychain,
+        RMC: Multicast<H>,
+    > BlockSignatureAggregator<'a, H, D, N, MK, RMC>
+where
+    RMC::Signed: Multisigned<'a, H, MK>,
+{
+    pub(crate) fn new(
+        network: N,
+        rmc: RMC,
+        messages_for_rmc: mpsc::UnboundedSender<D>,
+        messages_from_rmc: mpsc::UnboundedReceiver<D>,
+        metrics: Option<Metrics<H>>,
+    ) -> Self {
+        BlockSignatureAggregator {
+            messages_for_rmc,
+            messages_from_rmc,
+            signatures: HashMap::new(),
+            hash_queue: VecDeque::new(),
+            network,
+            multicast: rmc,
+            last_hash_placed: false,
+            started_hashes: HashSet::new(),
+            metrics,
+            marker: PhantomData,
+        }
+    }
+
+    pub(crate) async fn start_aggregation(&mut self, hash: H) {
+        debug!(target: "aleph-aggregator", "Started aggregation for block hash {:?}", hash);
+        if !self.started_hashes.insert(hash) {
+            debug!(target: "aleph-aggregator", "Aggregation already started for block hash {:?}, exiting.", hash);
+            return;
+        }
+        if let Some(metrics) = &self.metrics {
+            metrics.report_block(hash, std::time::Instant::now(), Checkpoint::Aggregating);
+        }
+        self.hash_queue.push_back(hash);
+        self.multicast
+            .start_multicast(SignableHash::new(hash))
+            .await;
+    }
+
+    pub(crate) fn notify_last_hash(&mut self) {
+        self.last_hash_placed = true;
+    }
+
+    pub(crate) async fn wait_for_next_signature(&mut self) -> Result<(), AggregatorError> {
```

Comment:
this one doesn't have to be exposed with `pub(crate)`, does it? or are you going to test it directly outside this module?

Reply:
At least for now I'm not, so I'll change it, thanks